### PR TITLE
@craigspaeth User restructure (do not merge)

### PR DESCRIPTION
**Pull request by kanaabe**
This is definitely not ready to merge yet (more tests and edge cases). When ready, I'll hit merge around July 18, test on staging for the day, and deploy to production on July 19.

Closes https://github.com/artsy/positron/issues/743 https://github.com/artsy/positron/issues/773

In browsing Writer, you'll now have a `current_channel` field which keeps track of the channel you are logged into.

### New Features
**Channel switching** -- instead of impersonating a user, you are now granted access to channels via http://writer.artsy.net/settings/channels or http://admin.artsy.net for partner access. Admins will be able to see all partner channels.
**Default publishing for Partners** -- when publishing in a partner channel, your post will automatically be added to the Partner's profile page.
**User refreshing** -- an outdated user will refresh itself in the DB and ask a user to log in again.
**Author renaming** -- instead of changing the ownership of an article, the author can be renamed.
**Unauthorized error page** -- a user who isn't part of a channel, partner, or admin will see a 403 page.

### CODE REVIEW

File: api/apps/articles/model/index.coffee [hunk 1]
```diff
@@ -48,16 +48,19 @@ Q = require 'bluebird-q'
...
-            if publishing
-              onPublish article, author, sanitizeAndSave(callback)
-            else if not publishing and not article.slugs?.length > 0
-              generateSlugs article, author, sanitizeAndSave(callback)
-            else
-              sanitizeAndSave(callback)(null, article)
+          publishing = (input.published and not article.published) or (input.scheduled_publish_at and not article.published)
+          article = _.extend article, _.omit(input, 'sections'), updated_at: new Date
+          if input.sections and input.sections.length is 0
+            article.sections = []
```
Comment (craigspaeth) [inline api/apps/articles/model/index.coffee]: Could we just use `.default([])` in [the schema](https://github.com/artsy/positron/blob/master/api/apps/articles/model/schema.coffee#L108) instead? Seems like that would allow you to do `article = _.extend article, input, updated_at: new Date` without the `omit`

File: api/apps/articles/model/schema.coffee [hunk 2]
```diff
@@ -49,6 +49,9 @@ denormalizedArtwork = (->
 @inputSchema = (->
   id: @objectId()
   author_id: @objectId().required()
+  author: @object().keys
```
Comment (craigspaeth) [inline api/apps/articles/model/schema.coffee]: 👍  makes sense given the use case

> instead of changing the ownership of an article, the author can be renamed

File: api/apps/channels/model.coffee [hunk 3]
```diff
@@ -29,20 +29,25 @@ querySchema = (->
...
+  q: @string()
 ).call Joi
 
 #
 # Retrieval
 #
 @find = (id, callback) ->
-  query = if ObjectId.isValid(id) then { _id: ObjectId(id) }
-  db.channels.findOne query, callback
+  return callback 'Invalid channel id' unless ObjectId.isValid(id)
```
Comment (craigspaeth) [inline api/apps/channels/model.coffee]: This [should be wrapped](http://www.devthought.com/2011/12/22/a-string-is-not-an-error/) in an error object `new Error('Invalid channel id')`.

File: api/apps/channels/model.coffee [hunk 4]
```diff
@@ -29,20 +29,25 @@ querySchema = (->
...
+  query = { _id: ObjectId(id) }
+  db.channels.findOne query, (err, channel) ->
+    return callback 'No channel found' unless channel
```
Comment (craigspaeth) [inline api/apps/channels/model.coffee]: Ditto.

File: api/apps/users/model.coffee [hunk 5]
```diff
@@ -40,16 +40,13 @@ bcrypt = require 'bcrypt'
...
-  db.users.findOne { _id: ObjectId(id) }, (err, user) ->
-    return callback err if err
-    return callback null, user if user
-    request.get("#{ARTSY_URL}/api/v1/user/#{id}")
-      .set('X-Access-Token': accessToken)
-      .end (err, user) ->
-        return callback err if err
-        save user.body, accessToken, callback
+
+@resave = (accessToken, callback) ->
```
Comment (craigspaeth) [inline api/apps/users/model.coffee]: 👌

File: client/apps/edit/components/admin/index.coffee [hunk 6]
```diff
@@ -380,6 +376,7 @@ module.exports = class EditAdmin extends Backbone.View
         @article.save super_article: superArticle
 
   setupSuperArticleAutocomplete: ->
+    return unless @channel.hasFeature 'superArticle'
```
Comment (craigspaeth) [inline client/apps/edit/components/admin/index.coffee]: 👍

File: client/apps/switch_channel/index.coffee [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+app = module.exports = express()
+
+app.get '/switch_channel/:id', switchChannel = (req, res, next) ->
+  return next() unless _.contains(req.user.get('channel_ids'), req.params.id) or
+    _.contains(req.user.get('partner_ids'), req.params.id) or
+    req.user.get('type') is 'Admin'
+  new Channel(id: req.params.id).fetchChannelOrPartner
+    success: (channel) =>
+      req.user.set 'current_channel', channel.denormalized()
```
Comment (craigspaeth) [inline client/apps/switch_channel/index.coffee]: 👍

File: client/components/layout/channel_autocomplete.coffee [hunk 8]
```diff
@@ -0,0 +1,81 @@
+{ defer } = require 'underscore'
+Modal = require 'simple-modal'
+sd = require('sharify').data
+async = require 'async'
+request = require 'superagent'
+User = require '../../models/user.coffee'
+
+module.exports.init = ->
+  $('#layout-sidebar-switch-channel').click =>
```
Comment (craigspaeth) [inline client/components/layout/channel_autocomplete.coffee]: This one big `init` function works fine for now—but I think ideally this gets refactored into a Backbone view for the sake of testability/modularity.

File: client/components/layout/client.coffee [hunk 9]
```diff
@@ -22,42 +22,23 @@ module.exports.init = ->
...
+  channelAutocomplete.init()
 
-# Replace broken profile icon
-imgLoad = imagesLoaded('#layout-sidebar-profile img')
-imgLoad.on 'fail', ->
-  $('#layout-sidebar-profile img').attr(
-    'src'
-    "/images/layout_missing_user.png"
-  )
+  # Replace broken profile icon
```
Comment (craigspaeth) [inline client/components/layout/client.coffee]: This isn't needed anymore with the profile icon being a generic svg right?

image

File: client/components/layout/client.coffee [hunk 10]
```diff
@@ -22,42 +22,23 @@ module.exports.init = ->
...
-  $('#layout-sidebar-container').toggleClass('is-active')
-
-initAnalyitcs = ->
-  if sd.USER
-    analytics.identify sd.USER.id,
-      email: sd.USER.email
-      name: sd.USER.name
-  analytics.page()
+  # Ensure a fresh user
+  @user.isOutdated (outdated) =>
```
Comment (craigspaeth) [inline client/components/layout/client.coffee]: 👌

File: client/lib/setup/auth.coffee [hunk 11]
```diff
@@ -35,18 +43,26 @@ logoutOldSchema = (req, res, next) ->
 requireLogin = (req, res, next) ->
   if req.user? then next() else res.redirect '/login'
 
+requireChannel = (req, res, next) ->
+  if req.user?.get('current_channel') then next() else req.logout()
```
Comment (craigspaeth) [inline client/lib/setup/auth.coffee]: 👍

File: client/models/channel.coffee [hunk 12]
```diff
@@ -1,29 +1,118 @@
...
-        sections: ['text', 'artworks', 'images', 'video', 'callout']
-        associations: ['artworks', 'artists', 'shows', 'fairs', 'partners','auctions']
-      }
-      when 'partner' then {
-        features: []
-        sections: ['text','artworks', 'images','video']
-        associations: []
-      }
\ No newline at end of file
+  hasFeature: (feature) ->
```
Comment (craigspaeth) [inline client/models/channel.coffee]: 👍  I like how this method reduces the feature sets for each channel to lists.

File: client/models/channel.coffee [hunk 13]
```diff
@@ -1,29 +1,118 @@
...
+        'artists'
+        'shows'
+        'fairs'
+        'partners'
+        'auctions'
+      ], association
+    else if type is 'partner'
+       false
+
+  fetchChannelOrPartner: (options) ->
```
Comment (craigspaeth) [inline client/models/channel.coffee]: Good work extracting these model methods!

File: client/models/user.coffee [hunk 14]
```diff
@@ -1,15 +1,62 @@
...
+  hasPartner: (id) ->
+    @isAdmin() or _.contains @get('partner_channel_ids'), id
+
+  hasArticleAccess: (article) ->
+    if article.get('channel_id')
+      @hasChannel article.get('channel_id')
+    else
+      @hasPartner article.get('partner_channel_id')
+
+  isOutdated: (callback) ->
```
Comment (craigspaeth) [inline client/models/user.coffee]: Nice!

**Comment by craigspaeth** [comment]
Looking good! Overall the architecture is sound, and I appreciate the refactors being conscious of when it makes sense to move to the API/model layer. I have some small comments, but generally this is on the right track. 👏  well done!

**Comment by kanaabe** [comment]
Thanks and sorry I didn't leave comments beforehand -- I'm going to do a few more passes and then I'll let you know when it's ready for a final review.

Definitely noted the `Error` handling, default values, and extracting the autocomplete to a view. I started with trying to make a generic `AutocompleteMultipleSource` view, but this case was pretty complicated with the local + remote data sources ha. Anyway I'll just make it a `AutocompleteChannel` view.

File: api/apps/articles/model/index.coffee (hunk 1 above)
Comment (kanaabe) [inline api/apps/articles/model/index.coffee]: Ah I believe this came about because of : https://github.com/artsy/positron/pull/689

File: api/apps/articles/model/retrieve.coffee [hunk 15]
```diff
@@ -27,14 +27,18 @@ moment = require 'moment'
...
-    query.partner_channel_id = ObjectId input.partner_channel_id if input.partner_channel_id
 
     # Convert query for super article for article
     query['super_article.related_articles']= ObjectId(input.super_article_for) if input.super_article_for
 
     # Only add the $or array for queries that require it (blank $or array causes problems)
-    query.$or ?= [] if input.artist_id or input.all_by_author or input.has_video
+    query.$or ?= [] if input.artist_id or input.all_by_author or input.has_video or input.channel_id
+
+    # Convert query for channel_id to include partner_channel_id
```
Comment (craigspaeth) [inline api/apps/articles/model/retrieve.coffee]: 👍

File: api/apps/articles/routes.coffee [hunk 16]
```diff
@@ -1,43 +1,56 @@
...
 
 # GET /api/articles
 @index = (req, res, next) ->
-  if req.query.published isnt 'true' and (not req.query.author_id? or
-     req.query.author_id isnt req.user?._id.toString()) and
-     req.user?.type isnt 'Admin'
-    return res.err 401,
-      'Must pass author_id=me to view unpublished articles. Or pass ' +
-      'published=true to only view published articles.'
+  return res.err(401, 'Must pass channel_id to view unpublished articles. Or pass ' +
```
Comment (craigspaeth) [inline api/apps/articles/routes.coffee]: Nice! The early returns definitely help following the auth logic here. Small style note: it might be easier to read by putting the `if`s first, and/or pulling the messages out into variables e.g. `res.err(401, messages.requiredParams) if ...`.

File: api/apps/articles/routes.coffee [hunk 17]
```diff
@@ -1,43 +1,56 @@
...
+  return res.send present req.article if req.article.published
+
+  if User.hasChannelAccess req.user, (req.article.channel_id or req.article.partner_channel_id)
     res.send present req.article
   else
     res.err 404, 'Article not found.'
 
 # POST /api/articles
 @create = (req, res, next) ->
+  unless User.hasChannelAccess req.user, (req.body.channel_id or req.body.partner_channel_id)
```
Comment (craigspaeth) [inline api/apps/articles/routes.coffee]: I think it would make sense to [DRY](https://www.wikiwand.com/en/Don't_repeat_yourself) this up by moving it to the [find](https://github.com/kanaabe/positron/blob/f24ec2c3018d825b2bf4f18f68bd8a7171b50c73/api/apps/articles/routes.coffee#L66) middleware. Something after [this line](https://github.com/kanaabe/positron/blob/f24ec2c3018d825b2bf4f18f68bd8a7171b50c73/api/apps/articles/routes.coffee#L69) like....

````coffeescript
channelID = (
  req.article.channel_id or
  req.article.partner_channel_id or
  req.body.channel_id or
  req.body.partner_channel_id
)
res.err(401, 'Unauthorized) unless User.hasChannelAccess req.user, channelID
````

File: api/apps/users/model.coffee [hunk 18]
```diff
@@ -78,19 +75,27 @@ save = (user, accessToken, callback) ->
     }, callback
 
 #
+# Utility
+#
+@hasChannelAccess = (user, channel_id) ->
```
Comment (craigspaeth) [inline api/apps/users/model.coffee]: 👍  nice extraction of logic here— a little much on the newline whitespace for my taste, but I don't claim to have much of a refined palette anyway.

File: api/apps/users/test/integration.coffee [hunk 19]
```diff
@@ -15,26 +15,38 @@ describe 'user endpoints', ->
...
-            done()
+  it 'returns yourself', (done) ->
+    request
+      .get("http://localhost:5000/users/me")
+      .set('X-Access-Token': @user.access_token)
+      .end (err, res) =>
+        res.body.name.should.equal @user.name
+        done()
+
+# describe 'GET /api/users/me/refresh', ->
```
Comment (craigspaeth) [inline api/apps/users/test/integration.coffee]: Oops? Should probably uncomment these or just delete them.

File: client/components/error/page.jade [hunk 20]
```diff
@@ -13,6 +13,10 @@ html(lang="en")
         h1#error-not-found Page not found, please check that you have the correct url.
           br
           a( href='/' ) Back to Artsy Writer
+      else if error.status == 403
+        h1#error-not-found This page is forbidden. Please check with support@artsy.net or your Artsy representative if you believe this is a mistake.
```
Comment (craigspaeth) [inline client/components/error/page.jade]: 🚫

File: client/test/models/user.coffee [hunk 21]
```diff
@@ -24,12 +27,177 @@ describe "User", ->
       @user = new User _.extend fixtures().users, type: 'User'
       @user.isAdmin().should.be.false()
 
-  describe '#isEditorialTeam', ->
+  describe '#hasChannel', ->
```
Comment (craigspaeth) [inline client/test/models/user.coffee]: Killer test coverage overall!!

**Comment by craigspaeth** [comment]
I have a [couple](https://github.com/artsy/positron/pull/772#discussion_r71399354) [of](https://github.com/artsy/positron/pull/772#discussion_r71400533) [mostly](https://github.com/artsy/positron/pull/772#discussion_r71401349) [style](https://github.com/artsy/positron/pull/772#discussion_r71401540) comments, but overall looks great—not blockers I can see that should prevent this from shipping. Feel free to merge when ready!

**Comment by kanaabe** [comment]
Thanks so much! Uncommented that test + did some cleanup. I'll work on getting the extracting the `User.hasChannelAccess` into its own middleware after. Merging!
